File: patterns/arrays/two-sum/two-sum.py
import random
import time

def two_sum_brute(nums, target):
    for i in range(len(nums)):
        for j in range(i +1, len(nums)):
            if nums[i] + nums[j] == target:
                return [i, j]
    return []

# Test input
nums = [random.randint(0, 10000) for _ in range(10000)]
target = nums[0] + nums[1]

# Benchmark Brute Force
start = time.perf_counter()
for _ in range(10):
    two_sum_brute(nums, target)
end = time.perf_counter()
print(f"Python Brute Force: {end - start:.6f} seconds")

File: patterns/arrays/two-sum/two-sum.ts
/**
 * DSA 1 – Two Sum
 * Patterns:
 *   1. Brute Force
 * Description: Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.
 * Constraints:
 *   - Exactly one solution exists
 *   - Cannot use the same element twice
 */

/** 
 * Pattern: Brute Force
 * Time: O(n^2)
 * Space: O(1)
 * Insight: Check every pair until sum equals target.
 */
function twoSumBrute(nums: number[], target: number): number[] {
    for (let i = 0; i < nums.length; i++) {
        for (let j = i + 1; j < nums.length; j++) {
            if (nums[i] + nums[j] === target) {
                return [i, j];
            }
        }
    }
    return [];
}


const n = Array.from({ length: 10000 }, () => Math.floor(Math.random() * 10000));
const t = n[0] + n[1];

// Benchmark Brute Force
console.time('TS Brute Force');
for (let i = 0; i < 10; i++) {
    twoSumBrute(n, t);
}
console.timeEnd('TS Brute Force');